### PR TITLE
Fix: hyperdrive over jumpdrive

### PR DESCRIPTION
**Pull request by RisingLeaf**
**Bugfix:** This PR addresses issue #7189
## Fix Details
Quick and dirty fix of hyperdrive over jumpdrive.
Refractoring comes latter.

### CODE REVIEW

File: source/Ship.cpp [hunk 1]
```diff
@@ -3158,11 +3158,13 @@ double Ship::JumpFuel(const System *destination) const
 
 	bool linked = currentSystem->Links().count(destination);
 	// Figure out what sort of jump we're making.
-	if(attributes.Get("hyperdrive") && linked)
+	double hyperFuelNeeded = HyperdriveFuel();
+	double jumpFuelNeeded =JumpDriveFuel((linked || currentSystem->JumpRange()) ? 0. : currentSystem->Position().Distance(destination->Position()));
```
Comment (thvk-net) [inline source/Ship.cpp:3162]:
```suggestion
	double jumpFuelNeeded = JumpDriveFuel((linked || currentSystem->JumpRange()) ? 0. : currentSystem->Position().Distance(destination->Position()));
```

File: source/Ship.cpp [hunk 2]
```diff
@@ -3158,11 +3158,13 @@ double Ship::JumpFuel(const System *destination) const
 
 	bool linked = currentSystem->Links().count(destination);
 	// Figure out what sort of jump we're making.
```
Comment (thvk-net) [inline source/Ship.cpp:3160]:
```suggestion
	// Figure out what sort of jump we're making. Compare how much fuel each method uses.
```

**Comment by Amazinite** [comment]
Note that I believe that this will only cause the amount of fuel consumed to be the minimum between the two jump methods when using a hyperspace link. Code for determining which jump method actually gets used and executing that jump method is done elsewhere (Ship::IsJumpReady, sections of Ship::Move, and AI::PrepareForHyperspace).

**Comment by seakintruth** [comment]
When I roll back to PR7208 they do jump. (running on Ubuntu)

**Comment by thvk-net** [comment]
> Note that I believe that this will only cause the amount of fuel consumed to be the minimum between the two jump methods when using a hyperspace link. Code for determining which jump method actually gets used and executing that jump method is done elsewhere (Ship::IsJumpReady, sections of Ship::Move, and AI::PrepareForHyperspace).

Oh, ok I will figure it out.

**Comment by thvk-net** [comment]
> 



> Note that I believe that this will only cause the amount of fuel consumed to be the minimum between the two jump methods when using a hyperspace link. Code for determining which jump method actually gets used and executing that jump method is done elsewhere (Ship::IsJumpReady, sections of Ship::Move, and AI::PrepareForHyperspace).

I added the same validation method to the two other functions

**Review by quyykk — COMMENTED** [review]
Nice! Although the couple of lines of duplicated code definitely belong in a new function.

**Review by Amazinite — CHANGES_REQUESTED** [review]
There's a lot of duplicate calculations being done here. For example, JumpDriveCheaper calls JumpDriveFuel and HyperdriveFuel, yet in Ship::JumpFuel you're calling JumpDriveCheaper and then calling JumpDriveFuel or HyperdriveFuel again, and in AI::PrepareForHyperspace you're calling JumpDriveCheaper and then checking if the ship has a jump drive, when JumpDriveCheaper is already calling other functions that figure that out.

I think that what you should do is have a method that returns the jump method and its cost at the same time: `pair<bool, double> Ship::JumpMethod(const System *destination)`
If no jump drive exists, the bool in the pair would always return false. We could also return an enum instead of a bool to be more transparent with what the bool represents, but I don't see that as entirely necessary.

File: source/AI.cpp [hunk 3]
```diff
@@ -1929,7 +1929,9 @@ void AI::PrepareForHyperspace(Ship &ship, Command &command)
 	if(!hasHyperdrive && !hasJumpDrive)
 		return;
 
-	bool isJump = !hasHyperdrive || !ship.GetSystem()->Links().count(ship.GetTargetSystem());
+	
+	bool isJump= ship.JumpDriveCheaper(ship.GetTargetSystem()) && ship.Attributes().Get("jump drive");
```
Comment (tibetiroka) [inline source/AI.cpp:1933]:
```suggestion
	bool isJump = ship.JumpDriveCheaper(ship.GetTargetSystem()) && ship.Attributes().Get("jump drive");
```

File: source/Ship.cpp [hunk 4]
```diff
@@ -3258,6 +3263,24 @@ double Ship::JumpDriveFuel(double jumpDistance) const
 
 
 
+bool Ship::JumpDriveCheaper(const System *destination) const
+{
+	bool linked = currentSystem->Links().count(destination);
+	double hyperFuelNeeded = HyperdriveFuel();
+	double jumpFuelNeeded = JumpDriveFuel((linked || currentSystem->JumpRange()) ? 0. : currentSystem->Position().Distance(destination->Position()));
```
Comment (tibetiroka) [inline source/Ship.cpp:3270]:
```suggestion
	double jumpFuelNeeded = JumpDriveFuel((linked || currentSystem->JumpRange())
			? 0. : currentSystem->Position().Distance(destination->Position()));
```
Lines shound't be longer than 120 characters. The exact position of the wrap is somewhat arbitrary.

**Review by tibetiroka — CHANGES_REQUESTED** [review]
Some minor style fixes.

**Comment by RisingLeaf** [comment]
> There's a lot of duplicate calculations being done here. For example, JumpDriveCheaper calls JumpDriveFuel and HyperdriveFuel, yet in Ship::JumpFuel you're calling JumpDriveCheaper and then calling JumpDriveFuel or HyperdriveFuel again, and in AI::PrepareForHyperspace you're calling JumpDriveCheaper and then checking if the ship has a jump drive, when JumpDriveCheaper is already calling other functions that figure that out.
> 
> I think that something you could do to avoid this is have a method that returns the jump method and its cost at the same time: `pair<bool, double> Ship::JumpMethod(const System *destination)` If no jump drive exists, the bool in the pair would always return false. We could also return an enum instead of a bool to be more transparent with what the bool represents, but I don't see that as entirely necessary. The double in the pair would then be the jump cost of the method being used. If no jump method is available then the jump cost is 0.

done

File: source/AI.cpp [hunk 5]
```diff
@@ -1929,7 +1929,8 @@ void AI::PrepareForHyperspace(Ship &ship, Command &command)
 	if(!hasHyperdrive && !hasJumpDrive)
 		return;
 
-	bool isJump = !hasHyperdrive || !ship.GetSystem()->Links().count(ship.GetTargetSystem());
+	std::pair<Ship::JumpType, double> jumpUsed = ship.JumpDriveCheaper(ship.GetTargetSystem());
+	bool isJump = jumpUsed.first == Ship::JUMPDRIVE;
```
Comment (RisingLeaf) [inline source/AI.cpp:1933]:
```suggestion
	bool isJump = jumpUsed.first == JUMPDRIVE;
```

**Comment by RisingLeaf** [comment]
> I just tested this PR and installed a Hyperdrive and a Jump Drive (while increasing the fuel required on the Hyperdrive) and .... I didn't jump. The ship just stands there, not doing anything lmao.

fixed

File: source/Ship.h [hunk 6]
```diff
@@ -310,6 +310,8 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	// Get the cost of making a jump of the given type (if possible).
 	double HyperdriveFuel() const;
 	double JumpDriveFuel(double jumpDistance = 0.) const;
+	enum JumpType{ HYPERDRIVE, JUMPDRIVE, NONE };
```
Comment (quyykk) [inline source/Ship.h:313]: Can you:

- Make this an `enum class`
- Move it up with the other class definitions
- Not caps the values

thx

File: source/Ship.h [hunk 7]
```diff
@@ -310,6 +310,8 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
...
+	std::pair<JumpType, double> JumpDriveCheaper(const System *destination) const;
```
Comment (quyykk) [inline source/Ship.h:314]: This name isn't really good. How about something like `GetCheapestDrive` or something else.

File: source/Ship.cpp [hunk 8]
```diff
@@ -3258,6 +3249,29 @@ double Ship::JumpDriveFuel(double jumpDistance) const
...
+std::pair<Ship::JumpType, double> Ship::JumpDriveCheaper(const System *destination) const
+{
+	bool linked = currentSystem->Links().count(destination);
+	double hyperFuelNeeded = HyperdriveFuel();
+	double jumpFuelNeeded = JumpDriveFuel((linked || currentSystem->JumpRange())
+			? 0. : currentSystem->Position().Distance(destination->Position()));
+	if(linked)
+	{
+		if(hyperFuelNeeded > jumpFuelNeeded && attributes.Get("jump drive"))
+			return std::make_pair(JUMPDRIVE, jumpFuelNeeded);
```
Comment (quyykk) [inline source/Ship.cpp:3261]: Drop the `std::` prefix in .cpp files.

**Review by quyykk — COMMENTED** [review]
Nice, thanks!

File: source/Ship.cpp (hunk 8 above)
Comment (RisingLeaf) [reply source/Ship.cpp]: done

File: source/Ship.h (hunk 7 above)
Comment (RisingLeaf) [reply source/Ship.h]: done

File: source/Ship.h (hunk 6 above)
Comment (RisingLeaf) [reply source/Ship.h]: done

File: source/Ship.cpp [hunk 9]
```diff
@@ -3258,6 +3249,29 @@ double Ship::JumpDriveFuel(double jumpDistance) const
...
+			return make_pair(JumpType::JumpDrive, jumpFuelNeeded);
+		else if(attributes.Get("hyperdrive"))
+			return make_pair(JumpType::HyperDrive, hyperFuelNeeded);
+		else
+			return make_pair(JumpType::None, 0.0);
+	}
+	else if(attributes.Get("jump drive"))
+		return make_pair(JumpType::JumpDrive, jumpFuelNeeded);
+	else
+		return make_pair(JumpType::None, 0.0);
```
Comment (Amazinite) [inline source/Ship.cpp:3270]: Simplifying this.
```suggestion
	if(linked && attributes.Get("hyperdrive") && hyperFuelNeeded <= jumpFuelNeeded)
		return make_pair(JumpType::Hyperdrive, hyperFuelNeeded);
	else if(attributes.Get("jump drive"))
		return make_pair(JumpType::JumpDrive, jumpFuelNeeded);
	else
		return make_pair(JumpType::None, 0.0);
```

**Review by Amazinite — CHANGES_REQUESTED** [review]
Ship::HyperdriveFuel calls Ship::JumpDriveFuel if you don't have a hyperdrive. That may have been necessary in the past, but as far as I can tell right now, HyperdriveFuel is always called alongside JumpDriveFuel, so HyperdriveFuel should probably just return 0 if you don't have a hyperdrive instead of effectively calling JumpDriveFuel twice.

File: source/Ship.cpp [hunk 10]
```diff
@@ -2804,7 +2805,7 @@ bool Ship::IsReadyToJump(bool waitingIsReady) const
 		return false;
```
Comment (Amazinite) [inline source/Ship.cpp:2805]: I can't comment high enough in the preview, but there's a line just above here that calls JumpFuel, which just ends up calling GetCheapestJumpType, so this code should look something like this:
```c++
pair<Ship::JumpType, double> jumpUsed = GetCheapestJumpType(targetSystem);
double fuelCost = jumpUsed.second;
if(!fuelCost || fuel < fuelCost)
	return false;

Point direction = targetSystem->Position() - currentSystem->Position();
bool isJump = (jumpUsed.first == JumpType::JumpDrive);
...
```

File: source/Ship.cpp [hunk 11]
```diff
@@ -3237,7 +3227,7 @@ double Ship::HyperdriveFuel() const
 {
 	// Don't bother searching through the outfits if there is no hyperdrive.
 	if(!attributes.Get("hyperdrive"))
-		return JumpDriveFuel();
+		return 0.0;
```
Comment (Amazinite) [inline source/Ship.cpp:3230]:
```suggestion
		return 0.;
```

File: source/Ship.cpp (hunk 10 above)
Comment (RisingLeaf) [reply source/Ship.cpp:2806]: done

**Comment by RisingLeaf** [comment]
I dont know what went wrong in the checks but I dont think its my fault

**Comment by Amazinite** [comment]
Ships are trying to refuel when they don't need to.

**Comment by RisingLeaf** [comment]
> Ships are trying to refuel when they don't need to. image

I will check this out, thanks

File: source/Ship.cpp [hunk 12]
```diff
@@ -3258,6 +3249,22 @@ double Ship::JumpDriveFuel(double jumpDistance) const
 
 
 
+pair<Ship::JumpType, double> Ship::GetCheapestJumpType(const System *destination) const
+{
+	bool linked = currentSystem->Links().count(destination);
+	double hyperFuelNeeded = HyperdriveFuel();
+	double jumpFuelNeeded = JumpDriveFuel((linked || currentSystem->JumpRange())
+			? 0. : currentSystem->Position().Distance(destination->Position()));
+	if(linked && attributes.Get("hyperdrive") && hyperFuelNeeded <= jumpFuelNeeded)
```
Comment (Amazinite) [inline source/Ship.cpp:3258]:
```suggestion
	if(linked && attributes.Get("hyperdrive") && (!jumpFuelNeeded || hyperFuelNeeded <= jumpFuelNeeded))
```

Comment (RisingLeaf) [reply source/Ship.cpp]: Oh yes, I figured out myself.

**Review by Amazinite — APPROVED** [review]

**Pull request merged**